salt/engines/libvirt_events.py

```diff
@@ -0,0 +1,617 @@
+             ('refresh', None)),
+    'nodedev': (('lifecycle', None),
+                ('update', None)),
+    'secret': (('lifecycle', None),
+               ('value_changed', None))
+}
+
+
+def _compute_subprefix(attr):
+    '''
+    Get the part before the first '_' or the end of @attr including
+    the potential '_'
+    '''
+    return ''.join((attr.split('_')[0], '_' if len(attr.split('_')) > 1 else ''))
+
+
+def _get_libvirt_enum_string(prefix, value):
+    '''
+    Convert the libvirt enum integer value @value into a human readable string.
+    @prefix is the start of the libvirt attribute to look for.
+    '''
+    attributes = [attr[len(prefix):] for attr in libvirt.__dict__ if attr.startswith(prefix)]
+
+    # Filter out the values starting with a common base as they match another enum
+    prefixes = [_compute_subprefix(p) for p in attributes]
+    counts = {p: prefixes.count(p) for p in prefixes}
+    sub_prefixes = [p for p, count in counts.items() if count > 1]
+    filtered = [attr for attr in attributes if _compute_subprefix(attr) not in sub_prefixes]
+
+    for candidate in filtered:
+        if value == getattr(libvirt, ''.join((prefix, candidate))):
+            name = candidate.lower().replace('_', ' ')
+            return name
+    return 'unknown'
+
+
+def _get_domain_event_detail(event, detail):
+    '''
+    Convert @event and @detail numeric values into a tuple of human readable strings
+    '''
+    event_name = _get_libvirt_enum_string('VIR_DOMAIN_EVENT_', event)
+    if event_name == 'unknown':
+        return event_name, 'unknown'
+
+    prefix = 'VIR_DOMAIN_EVENT_{0}_'.format(event_name.upper())
+    detail_name = _get_libvirt_enum_string(prefix, detail)
+
+    return event_name, detail_name
+
+
+def salt_send_event(opaque, conn, data):
+    '''
+    Convenience function adding common data to the event and sending it
+    on the salt event bus.
+    '''
+    tag_prefix = opaque['prefix']
+    object_type = opaque['object']
+    event_type = opaque['event']
+
+    # Prepare the connection URI to fit in the tag
+    # qemu+ssh://user@host:1234/system -> qemu+ssh/user@host:1234/system
+    uri = urlparse(conn.getURI())
+    uri_tag = [uri.scheme]
+    if uri.netloc:
+        uri_tag.append(uri.netloc)
+    path = uri.path.strip('/')
+    if path:
+        uri_tag.append(path)
+    uri_str = "/".join(uri_tag)
+
+    # Append some common data
+    all_data = {
+        'uri': conn.getURI()
+    }
+    all_data.update(data)
+
+    tag = '/'.join((tag_prefix, uri_str, object_type, event_type))
+
+    # Actually send the event in salt
+    if __opts__.get('__role') == 'master':
+        salt.utils.event.get_master_event(
+            __opts__,
+            __opts__['sock_dir']).fire_event(tag, all_data)
+    else:
+        __salt__['event.send'](tag, all_data)
+
+
+def salt_send_domain_event(opaque, conn, domain, event, event_data):
+    '''
+    Helper function send a salt event for a libvirt domain
+    '''
```

Comment:
Same comment here about docs.

Reply:
Rather than commenting in many places repeatedly, I'll just say this here: Any public facing function needs docs for all args/kwargs and some examples of usage. If you could update all of those below, that would be great.

Reply:
All these helper functions aren't intended to be used by external code, I'ld rather hide them, but adding those comments can't harm

Reply:
@cbosdo even hidden usually should at least document what it does. But yes, I would underscore them in front, since they are all internal.